File: Sources/mGBAEclipseCore/mGBAEclipseCore.cpp
#include "mGBAEclipseCore.hpp"

#include <mgba-util/memory.h>
#include <mgba-util/vfs.h>
#include <mgba/core/blip_buf.h>
#include <mgba/core/cheats.h>
#include <mgba/core/core.h>
#include <mgba/core/serialize.h>
#include <mgba/core/version.h>
#include <mgba/gba/core.h>
#include <mgba/internal/gba/audio.h>
#include <mgba/internal/gba/cheats.h>
#include <mgba/internal/gba/gba.h>
#include <mgba/internal/gba/input.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <unordered_map>
#include <sstream>
#include <algorithm>

void _log(
    struct mLogger* logger,
    int category,
    enum mLogLevel level,
    const char* format,
    va_list args)
{
}

static struct mLogger noopLogCtx = { .log = _log };

namespace eclipsecore {

struct CoreContext {
    const EKCoreCallbacks *callbacks;
    struct mCore *core;
    
    int16_t *samples;
    
    uint8_t *videoBuffer;
    uint32_t videoHeight;
    uint32_t videoWidth;
    uint32_t videoBufferSize;
    
    std::unordered_map<std::string, struct mCheatSet *> cheatSets;
};

#pragma mark - Consts

static const size_t audioChannelCount = 2;
static const size_t audioSampleRate = 32768;
static const size_t sampleCount = 1024;

const size_t audioBufferSize = sampleCount * audioSampleRate;
const EKCoreSetting settings[] = {};

const char *characterSetHexadeciaml = "1234567890ABCDEFabcdef";

EKCheatFormat cheatFormats[] = {
    {
        .id = "gba.gameshark",
        .displayName = "GameShark",
        .characterSet = EKCoreCheatCharacterSetHexadecimal,
        .format = "xxxxxxxx xxxxxxxx",
    },
    {
        .id = "gba.actionreplay-max",
        .displayName = "ActionReplay MAX",
        .characterSet = EKCoreCheatCharacterSetHexadecimal,
        .format = "xxxxxxxx xxxxxxxx"
    },
    {
        .id = "gba.codebreaker",
        .displayName = "CodeBreaker",
        .characterSet = EKCoreCheatCharacterSetHexadecimal,
        .format = "xxxxxxxx xxxx",
    }
};

#pragma mark - Setup

void deallocate(void* data)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);

    mCoreConfigDeinit(&ctx->core->config);
    ctx->core->deinit(ctx->core);
    free(ctx->samples);
    free(ctx);
}

EKCoreAudioFormat getAudioFormat(void* data)
{
    return {
        .commonFormat = EKCoreCommonAudioFormatPcmInt16,
        .sampleRate = audioSampleRate,
        .channelCount = audioChannelCount,
    };
}

EKCoreVideoFormat getVideoFormat(void* data)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);

    return {
        .pixelFormat = EKCoreVideoPixelFormatRgba8Unorm,
        .renderingType = EKCoreVideoRenderingTypeFrameBuffer,
        .width = ctx->videoWidth,
        .height = ctx->videoHeight,
    };
}

double getDesiredFrameRate(void* data)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    return ctx->core->frequency(ctx->core) / ctx->core->frameCycles(ctx->core);
}

bool canSetVideoPointer(void* data)
{
    return true;
}

uint8_t* getVideoPointer(void* data, uint8_t* preferredPointer)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    if (preferredPointer != nullptr) {
        ctx->videoBuffer = preferredPointer;
        ctx->core->setVideoBuffer(ctx->core, (color_t*)preferredPointer, ctx->videoWidth);
    }
    return ctx->videoBuffer;
}

#pragma mark - Lifecycle

bool start(void* data, const char* const romPath, const char* const savePath)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    ctx->core->opts.savegamePath = strdup(savePath);
    
    if (!mCoreLoadFile(ctx->core, romPath)) {
        return false;
    }

    mCoreLoadSaveFile(ctx->core, savePath, false);

    ctx->core->reset(ctx->core);

    return true;
}

void stop(void* data) { }

void restart(void* data)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    ctx->core->reset(ctx->core);
}

void play(void* data) { }
void pause(void* data) { }

void executeFrame(void* data, bool willRender)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);

    ctx->core->runFrame(ctx->core);

    blip_t* left = ctx->core->getAudioChannel(ctx->core, 0);
    blip_t* right = ctx->core->getAudioChannel(ctx->core, 1);

    int available = blip_samples_avail(left);
    int isOverLimit = (available > eclipsecore::sampleCount);
    available = isOverLimit * eclipsecore::sampleCount + (isOverLimit ^ 1) * available;

    blip_read_samples(left, ctx->samples, available, true);
    blip_read_samples(right, ctx->samples + 1, available, true);

    size_t interleavedAvailableBytes = available << 2;
    ctx->callbacks->writeAudio(ctx->callbacks->callbackContext, ctx->samples, interleavedAvailableBytes);
}

#pragma mark - Saves & Save States

bool save(void* data, const char* path)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    struct VFile* vf = VFileOpen(path, O_CREAT | O_TRUNC | O_RDWR);
    struct GBA *gba = static_cast<struct GBA *>(ctx->core->board);
    bool success = GBASavedataClone(&gba->memory.savedata, vf);
    return success;
}

void saveCallback(void* data)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    ctx->callbacks->didSave(ctx->callbacks->callbackContext);
}

bool saveState(void* data, const char* path)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    struct VFile* vf = VFileOpen(path, O_CREAT | O_TRUNC | O_RDWR);
    bool success = mCoreSaveStateNamed(ctx->core, vf, SAVESTATE_SAVEDATA);
    vf->close(vf);
    return success;
}

bool loadState(void* data, const char* path)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    struct VFile* vf = VFileOpen(path, O_RDONLY);
    bool success = mCoreLoadStateNamed(ctx->core, vf, SAVESTATE_SAVEDATA);
    vf->close(vf);
    return success;
}

uint8_t getMaxPlayers(void* data) { return 1; }

/// Notifies a core that a new player has connected.
bool playerConnected(void* data, uint8_t player) { return true; }

/// Notifies a core that a player has been disconnected.
void playerDisconnected(void* data, uint8_t player) { }

void playerSetInputs(void* data, uint8_t player, uint32_t inputs)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);

    // FIXME: see about implementing some GPIO-related inputs like gyro and maybe even the solar sensor
    
    uint32_t keys =
        ((inputs & EKInputFaceButtonRight) != 0) * (1 << GBA_KEY_A) |
        ((inputs & EKInputFaceButtonDown) != 0) * (1 << GBA_KEY_B) |
        ((inputs & EKInputShoulderLeft) != 0) * (1 << GBA_KEY_L) |
        ((inputs & EKInputShoulderRight) != 0) * (1 << GBA_KEY_R) |
        ((inputs & EKInputStartButton) != 0) * (1 << GBA_KEY_START) |
        ((inputs & EKInputSelectButton) != 0) * (1 << GBA_KEY_SELECT) |
        ((inputs & EKInputDpadUp) != 0) * (1 << GBA_KEY_UP) |
        ((inputs & EKInputDpadDown) != 0) * (1 << GBA_KEY_DOWN) |
        ((inputs & EKInputDpadLeft) != 0) * (1 << GBA_KEY_LEFT) |
        ((inputs & EKInputDpadRight) != 0) * (1 << GBA_KEY_RIGHT);
    
    ctx->core->setKeys(ctx->core, keys);
}

bool setCheat(void* data, const char* format, const char *code, bool enabled)
{
    CoreContext* ctx = static_cast<CoreContext*>(data);
    
    std::string codeStr = code;
    codeStr.erase(std::remove(codeStr.begin(), codeStr.end(), ' '), codeStr.end());
    
    std::stringstream hashKeyBuilder;
    hashKeyBuilder << format << '$' << codeStr;
    std::string hashKey = hashKeyBuilder.str();
    
    mCheatSet *set = ctx->cheatSets[hashKey];
    if (set != nullptr) {
        set->enabled = enabled;
        return;
    }

    struct mCheatDevice* device = ctx->core->cheatDevice(ctx->core);
    set = device->createSet(device, hashKey.c_str());

    size_t size = mCheatSetsSize(&device->cheats);
    if (size) {
        set->copyProperties(set, *mCheatSetsGetPointer(&device->cheats, size - 1));
    }
    
    size_t last = 0;
    size_t next = 0;
    while ((next = codeStr.find('\n', last)) != std::string::npos) {
        mCheatAddLine(set, codeStr.substr(last, next - last).c_str(), GBA_CHEAT_AUTODETECT);
        last = next + 1;
    }
    mCheatAddLine(set, codeStr.substr(last).c_str(), GBA_CHEAT_AUTODETECT);

    set->enabled = enabled;
    ctx->cheatSets[hashKey] = set;
    mCheatAddSet(device, set);
    
    return true;
}

void clearCheats(void *data) {
    CoreContext* ctx = static_cast<CoreContext*>(data);
    mCheatDeviceClear(ctx->core->cheatDevice(ctx->core));
    ctx->cheatSets.clear();
}
}

const EKCoreInfo mGBACoreInfo = {
    .id = "dev.magnetar.eclipseemu.mgba",
    .name = "mGBA",
    .developer = "Vicki Pfau",
    .version = "0.10.3",
    .sourceCodeUrl = "https://github.com/mgba-emu/mgba",
    .settings = {
        .version = 0,
        .itemsCount = sizeof(eclipsecore::settings) / sizeof(EKCoreSetting),
        .items = eclipsecore::settings,
    },
    .cheatFormatsCount = sizeof(eclipsecore::cheatFormats) / sizeof(EKCheatFormat),
    .cheatFormats = eclipsecore::cheatFormats,

    .setup = [](EKSystem system, const EKCoreCallbacks* callbacks) -> EKCore* {
        // do allocations

        eclipsecore::CoreContext* coreContext = static_cast<eclipsecore::CoreContext*>(malloc(sizeof(eclipsecore::CoreContext)));
        if (coreContext == nullptr) {
            return nullptr;
        }

        int16_t* audioBuffer = static_cast<int16_t*>(malloc(sizeof(int16_t) * eclipsecore::audioBufferSize));
        if (audioBuffer == nullptr) {
            free(coreContext);
            return nullptr;
        }

        EKCore* core = static_cast<EKCore*>(malloc(sizeof(EKCore)));
        if (core == nullptr) {
            free(coreContext);
            free(audioBuffer);
            return nullptr;
        }

        mLogSetDefaultLogger(&noopLogCtx);

        struct mCore* mgbaCore = GBACoreCreate();
        if (mgbaCore == nullptr) {
            free(coreContext);
            free(audioBuffer);
            free(core);
            return nullptr;
        }

        // setup mgba

        mCoreInitConfig(mgbaCore, nullptr);

        struct mCoreOptions mgbaCoreOptions = {
            .useBios = true
        };
        mCoreConfigLoadDefaults(&mgbaCore->config, &mgbaCoreOptions);
        if (mgbaCore->init(mgbaCore) != true) {
            free(coreContext);
            free(audioBuffer);
            free(core);
            mgbaCore->deinit(mgbaCore);
            return nullptr;
        }

        struct mCoreCallbacks mgbaCallbacks = {
            .context = (void*)coreContext,
            .savedataUpdated = eclipsecore::saveCallback,
        };
        mgbaCore->addCoreCallbacks(mgbaCore, &mgbaCallbacks);
        
        unsigned int videoWidth = 0;
        unsigned int videoHeight = 0;

        mgbaCore->desiredVideoDimensions(mgbaCore, &videoWidth, &videoHeight);
        mgbaCore->setAudioBufferSize(mgbaCore, eclipsecore::sampleCount);

        blip_t* left = mgbaCore->getAudioChannel(mgbaCore, 0);
        blip_t* right = mgbaCore->getAudioChannel(mgbaCore, 1);

        double clockRate = mgbaCore->frequency(mgbaCore);
        double fauxClock = GBAAudioCalculateRatio(1, 60, 1);

        blip_set_rates(left, clockRate, eclipsecore::audioSampleRate * fauxClock);
        blip_set_rates(right, clockRate, eclipsecore::audioSampleRate * fauxClock);

        // setup the core context

        coreContext->core = mgbaCore;
        coreContext->callbacks = callbacks;
        coreContext->samples = audioBuffer;
        coreContext->videoHeight = videoHeight;
        coreContext->videoWidth = videoWidth;
        coreContext->videoBufferSize = videoHeight * videoWidth * BYTES_PER_PIXEL;
        coreContext->cheatSets = std::unordered_map<std::string, struct mCheatSet *>();

        // setup the core

        core->data = coreContext;
        core->deallocate = eclipsecore::deallocate;
        core->getAudioFormat = eclipsecore::getAudioFormat;
        core->getVideoFormat = eclipsecore::getVideoFormat;
        core->getDesiredFrameRate = eclipsecore::getDesiredFrameRate;
        core->canSetVideoPointer = eclipsecore::canSetVideoPointer;
        core->getVideoPointer = eclipsecore::getVideoPointer;
        core->start = eclipsecore::start;
        core->stop = eclipsecore::stop;
        core->restart = eclipsecore::restart;
        core->play = eclipsecore::play;
        core->pause = eclipsecore::pause;
        core->executeFrame = eclipsecore::executeFrame;
        core->save = eclipsecore::save;
        core->saveState = eclipsecore::saveState;
        core->loadState = eclipsecore::loadState;
        core->getMaxPlayers = eclipsecore::getMaxPlayers;
        core->playerConnected = eclipsecore::playerConnected;
        core->playerDisconnected = eclipsecore::playerDisconnected;
        core->playerSetInputs = eclipsecore::playerSetInputs;
        core->setCheat = eclipsecore::setCheat;
        core->clearCheats = eclipsecore::clearCheats;

        return core;
    }
};

File: Sources/mGBAEclipseCore/mGBAEclipseCore.swift
import EclipseKit
import Foundation.NSURL
import mGBA

let audioChannelCount: UInt8 = 2;
let audioSampleRate: Double = 32768;
let sampleCount = 1024;

private extension URL {
	func filePath() -> String {
		if #available(iOS 16.0, *) {
			self.path(percentEncoded: false)
		} else {
			self.path
		}
	}
}

private func VFileOpen(path: URL, flags: Int32) -> UnsafeMutablePointer<VFile>? {
	VFileOpen(path.filePath(), flags)
}

internal extension CoreInputDelta {
	@inlinable
	func setInput(_ keys: inout UInt32, input: CoreInput, flag: UInt32, isPressed: Bool, useValue: Bool = true) {
		let isPressed: UInt32 = isPressed ? 1 : 0
		keys = self.input.contains(input) && useValue
			? (isPressed * (keys | flag)) + ((isPressed ^ 1) * (keys & ~flag))
			: keys
	}
}

public struct mGBAEclipseCoreSettings: CoreSettings {
    public static let currentVersion: Int16 = 0
    public static let descriptor: CoreSettingsDescriptor<mGBAEclipseCoreSettings> = .init(sections: [])

	public init() {}
    
    public static func decode(_ data: Data, version: Int16) throws -> mGBAEclipseCoreSettings {
        .init()
    }
    
    public func encode() throws -> Data {
        Data()
    }
}

private final class BridgeWrapper: CoreBridgeProtocol {
	public var bridge: any CoreBridgeProtocol

	init(bridge: consuming any CoreBridgeProtocol) {
		self.bridge = bridge
	}

	@inlinable
	func writeAudioSamples(samples: UnsafeRawBufferPointer) -> Int {
		self.bridge.writeAudioSamples(samples: samples)
	}
	
	@inlinable
	func didSave() {
		self.bridge.didSave()
	}
}

func handleSaveCallback(context: UnsafeMutableRawPointer?) {
	guard let context else { return }
	let wrapper = Unmanaged<BridgeWrapper>.fromOpaque(context)
	wrapper.takeUnretainedValue().didSave()
}

public final class mGBAEclipseCore: CoreProtocol {
	public typealias VideoRenderer = CoreFrameBufferVideoRenderer
	public typealias Settings = mGBAEclipseCoreSettings

	public enum Failure: LocalizedError {
		case allocatingCore
		case failedToLoadROM
		case failedToOpenFile
		case failedToSave
		case failedToLoadState
		case failedToSaveState
	}

	public static let id: String = "dev.magnetar.eclipseemu.mba"
	public static let name: String = "mGBA"
	public static let sourceCodeRepository: URL = .init(string: "https://github.com/mgba-emu/mgba")!
	public static let developer: String = "Vicki Pfau"
	public static let version: String = "0.10.5"
	public static let systems: Set<System> = [.gba]

	public static func features(for system: EclipseKit.System) -> EclipseKit.CoreFeatures {
		guard system == .gba else { return [] }
		return [.cheats, .saving, .saveStates]
	}
	
	public static func cheatFormats(for system: System) -> [CoreCheatFormat] {
		guard system == .gba else { return [] }
		return [
			.init(
				id: "gba.gameshark",
				name: "GameShark",
				charset: .hexadecimal,
				pattern: "xxxxxxxx xxxxxxxx"
			),
			.init(
				id: "gba.actionreplay-max",
				name: "ActionReplay MAX",
				charset: .hexadecimal,
				pattern: "xxxxxxxx xxxxxxxx"
			),
			.init(
				id: "gba.codebreaker",
				name: "CodeBreaker",
				charset: .hexadecimal,
				pattern: "xxxxxxxx xxxx"
			)
		]
	}
	
	public var bridge: any CoreBridgeProtocol
	public let settings: CoreResolvedSettings<mGBAEclipseCoreSettings>
	public let playerConnectionBehavior: CorePlayerConnectionBehavior = .linear
	public let maxPlayers: UInt8 = 1
	public var desiredFrameRate: Double = 60.0

	private let core: UnsafeMutablePointer<mCore>
	private let callbackContext: UnsafeMutableRawPointer
	private var cheatSets: [CoreCheat : UnsafeMutablePointer<mCheatSet>] = [:]

	private let videoHeight: UInt16
	private let videoWidth: UInt16
	private let videoBufferSize: UInt32
	private var videoBuffer: UnsafeMutableBufferPointer<UInt8>!
	private let samples: UnsafeMutableBufferPointer<Int16>!
	private var inputState: UInt32 = 0

	public init(
		system: System,
		settings: consuming CoreResolvedSettings<mGBAEclipseCoreSettings>,
		bridge: consuming any CoreBridgeProtocol
	) throws(mGBAEclipseCore.Failure) {
		setNoopLoggerAsGlobal();

		guard let core = GBACoreCreate() else {
			throw .allocatingCore
		}

		mCoreInitConfig(core, nil)
		var coreOptions = mCoreOptions()
		coreOptions.useBios = true

		mCoreConfigLoadDefaults(&core.pointee.config, &coreOptions)
		if (core.pointee.`init`(core) != true) {
			core.pointee.deinit(core)
			throw .allocatingCore
		}

		var videoWidth: UInt32 = 0
		var videoHeight: UInt32 = 0

		core.pointee.desiredVideoDimensions(core, &videoWidth, &videoHeight)
		core.pointee.setAudioBufferSize(core, sampleCount)

		let left = core.pointee.getAudioChannel(core, 0)
		let right = core.pointee.getAudioChannel(core, 1)
		let clockRate = core.pointee.frequency(core)
		let fauxClock = GBAAudioCalculateRatio(1, 60, 1)
		blip_set_rates(left, Double(clockRate), audioSampleRate * Double(fauxClock))
		blip_set_rates(right, Double(clockRate), audioSampleRate * Double(fauxClock))

		let wrapper = BridgeWrapper(bridge: bridge)
		callbackContext = Unmanaged.passRetained(wrapper).toOpaque()

		var coreCallbacks: mCoreCallbacks = .init()
		coreCallbacks.context = callbackContext
		coreCallbacks.savedataUpdated = handleSaveCallback

		core.pointee.addCoreCallbacks(core, &coreCallbacks)

		self.bridge = wrapper
		self.settings = settings
		self.core = core
		self.videoWidth = UInt16(videoWidth)
		self.videoHeight = UInt16(videoHeight)
		self.videoBufferSize = videoHeight * videoWidth * 4
		self.samples = .allocate(capacity: MemoryLayout<Int16>.stride * sampleCount * Int(audioSampleRate))
	}

	deinit {
		mCoreConfigDeinit(&core.pointee.config)
		core.pointee.deinit(core)
		self.samples.deallocate()
		_ = Unmanaged<BridgeWrapper>.fromOpaque(callbackContext).takeRetainedValue()
	}

	public func getVideoDescriptor() -> CoreVideoDescriptor {
		return .init(
			width: self.videoWidth,
			height: self.videoHeight,
			pixelFormat: .rgba8Unorm,
			frameBuffer: .assignable
		)
	}

	public func getAudioDescriptor() -> CoreAudioDescriptor {
		return .init(
			sampleRate: audioSampleRate,
			sampleFormat: .int16,
			channelCount: audioChannelCount
		)
	}
	
	public func setFrameBuffer(to pointer: UnsafeMutableBufferPointer<UInt8>) {
		self.videoBuffer = pointer
		self.videoBuffer.withMemoryRebound(to: color_t.self) { pointer in
			core.pointee.setVideoBuffer(core, pointer.baseAddress, Int(self.videoWidth))
		}
	}

	public func start(romPath: URL, savePath: URL) throws(mGBAEclipseCore.Failure) {
		let romPath = romPath.filePath()
		let savePath = savePath.filePath()
		core.pointee.opts.savegamePath = strdup(savePath)
		guard mCoreLoadFile(core, romPath) else {
			throw .failedToLoadROM
		}
		mCoreLoadSaveFile(core, savePath, false)
		core.pointee.reset(core)
	}

	public func stop() {}

	public func play() {}

	public func pause() {}

	public func reset(kind: CoreResetKind) {
		core.pointee.reset(core)
	}

	public func step(timestamp: CFAbsoluteTime, willRender: Bool) {
		core.pointee.runFrame(core)

		let left = core.pointee.getAudioChannel(core, 0)
		let right = core.pointee.getAudioChannel(core, 1)

		var available = blip_samples_avail(left)
		let isOverLimit: Int32 = available > sampleCount ? 1 : 0
		available = isOverLimit * Int32(sampleCount) + (isOverLimit ^ 1) * available

		blip_read_samples(left, samples.baseAddress, available, 1)
		blip_read_samples(right, samples.baseAddress?.advanced(by: 1), available, 1)

		let samples = UnsafeRawBufferPointer(start: samples.baseAddress, count: Int(available) << 2)
		bridge.writeAudioSamples(samples: samples)
	}

	public func save(to path: URL) async throws(mGBAEclipseCore.Failure) {
		guard let vf = VFileOpen(path: path, flags: O_CREAT | O_TRUNC | O_RDWR) else {
			throw .failedToOpenFile
		}
		defer { _ = vf.pointee.close(vf) }

		let gba = core.pointee.board.assumingMemoryBound(to: GBA.self)
		guard GBASavedataClone(&gba.pointee.memory.savedata, vf) else {
			throw .failedToSave
		}
	}

	public func saveState(to path: URL) throws(mGBAEclipseCore.Failure) {
		guard let vf = VFileOpen(path: path, flags: O_CREAT | O_TRUNC | O_RDWR) else {
			throw .failedToOpenFile
		}
		defer { _ = vf.pointee.close(vf) }

		guard mCoreSaveStateNamed(core, vf, SAVESTATE_SAVEDATA) else {
			throw .failedToLoadState
		}
	}

	public func loadState(from path: URL) throws(mGBAEclipseCore.Failure) {
		guard let vf = VFileOpen(path: path, flags: O_RDONLY) else {
			throw .failedToOpenFile
		}
		defer { _ = vf.pointee.close(vf) }

		guard mCoreLoadStateNamed(core, vf, SAVESTATE_SAVEDATA) else {
			throw .failedToLoadState
		}
	}

	public func playerConnected(to port: UInt8) {}

	public func playerDisconnected(from port: UInt8) {}

	public func writeInput(_ delta: CoreInputDelta, for player: UInt8) {
		var keys: UInt32 = self.inputState
		delta.setInput(&keys, input: .faceButtonRight, flag: 1 << GBA_KEY_A.rawValue, isPressed: delta.isPressed)
		delta.setInput(&keys, input: .faceButtonDown, flag: 1 << GBA_KEY_B.rawValue, isPressed: delta.isPressed)
		delta.setInput(&keys, input: .leftShoulder, flag: 1 << GBA_KEY_L.rawValue, isPressed: delta.isPressed)
		delta.setInput(&keys, input: .rightShoulder, flag: 1 << GBA_KEY_R.rawValue, isPressed: delta.isPressed)
		delta.setInput(&keys, input: .start, flag: 1 << GBA_KEY_START.rawValue, isPressed: delta.isPressed)
		delta.setInput(&keys, input: .select, flag: 1 << GBA_KEY_SELECT.rawValue, isPressed: delta.isPressed)
		delta.setInput(&keys, input: .select, flag: 1 << GBA_KEY_SELECT.rawValue, isPressed: delta.isPressed)
		delta.setInput(&keys, input: .dpad, flag: 1 << GBA_KEY_UP.rawValue, isPressed: delta.isUp, useValue: delta.useY)
		delta.setInput(&keys, input: .dpad, flag: 1 << GBA_KEY_DOWN.rawValue, isPressed: delta.isDown, useValue: delta.useY)
		delta.setInput(&keys, input: .dpad, flag: 1 << GBA_KEY_LEFT.rawValue, isPressed: delta.isLeft, useValue: delta.useX)
		delta.setInput(&keys, input: .dpad, flag: 1 << GBA_KEY_RIGHT.rawValue, isPressed: delta.isRight, useValue: delta.useX)
		self.inputState = keys
		core.pointee.setKeys(core, keys)
	}

	public func setCheat(cheat: CoreCheat, enabled: Bool) {
		if let set = cheatSets[cheat] {
			set.pointee.enabled = enabled
		}

		guard
			let device = core.pointee.cheatDevice(core),
			let set = device.pointee.createSet(device, "\(cheat.format)\(cheat.code)")
		else {
			return
		}

		let size = mCheatSetsSize(&device.pointee.cheats)
		if size > 0 {
			set.pointee.copyProperties(set, mCheatSetsGetPointer(&device.pointee.cheats, size - 1).pointee)
		}

		for line in cheat.code.trimmingCharacters(in: .whitespacesAndNewlines).split(separator: "\n") {
			_ = line.withCString { cString in
				mCheatAddLine(set, cString, Int32(GBA_CHEAT_AUTODETECT.rawValue))
			}
		}

		set.pointee.enabled = enabled
		cheatSets[cheat] = set
		mCheatAddSet(device, set)
	}

	public func clearCheats() {
		mCheatDeviceClear(core.pointee.cheatDevice(core))
		cheatSets.removeAll()
	}
}
